src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
+    // may have some unused bits. But the shards themselves are always a multiple of
+    // the word size. In practice the tail end of the index will be accessed less
+    // frequently than the head/middle anyway.
+    //
+    // Each shard is 8 cache lines because the CPU line fill buffer can fetch 10 lines in parallel. And 8 is fast for division when computing the shard of a block.
+    // Since the shard is scanned sequentially, the prefetching amortizes the cost of the single
+    // cache miss. It also reduces the size of the index.
+    //
+    // e.g. 10TiB disk ÷ 64KiB/block ÷ 512*8 blocks/shard ÷ 8 shards/byte = 5120B index
+    const shard_cache_lines = 8;
+    const shard_size = shard_cache_lines * config.cache_line_size * @bitSizeOf(u8);
+    comptime {
+        assert(shard_size == 4096);
+        assert(@bitSizeOf(MaskInt) == 64);
+    }
+
+    pub fn init(allocator: *mem.Allocator, blocks_count: usize) !BlockFreeSet {
+        assert(shard_size <= blocks_count);
+        assert(blocks_count % shard_size == 0);
+        assert(blocks_count % @bitSizeOf(usize) == 0);
+
+        // Round up to ensure that every block bit is covered by the index.
+        const shards_count = div_ceil(blocks_count, shard_size);
+        var index = try DynamicBitSetUnmanaged.initFull(shards_count, allocator);
+        errdefer index.deinit(allocator);
+
+        var blocks = try DynamicBitSetUnmanaged.initFull(blocks_count, allocator);
+        errdefer blocks.deinit(allocator);
 
-    pub fn init(allocator: *mem.Allocator, count: usize) !BlockFreeSet {
         return BlockFreeSet{
-            .free = try std.bit_set.DynamicBitSetUnmanaged.initFull(count, allocator),
+            .index = index,
+            .blocks = blocks,
+            .recent = .{},
         };
     }
 
     pub fn deinit(set: *BlockFreeSet, allocator: *mem.Allocator) void {
-        set.free.deinit(allocator);
+        set.index.deinit(allocator);
+        set.blocks.deinit(allocator);
     }
 
-    // TODO consider "caching" the first set bit to speed up subsequent acquire() calls
-    pub fn acquire(set: *BlockFreeSet) u64 {
-        // TODO: To ensure this "unreachable" is never reached, the leader must reject
-        // new requests when storage space is too low to fulfill them.
-        const bit = set.free.findFirstSet() orelse unreachable;
-        set.free.unset(bit);
-        const address = bit + 1;
+    /// Marks a free block as allocated, and returns the address. Panics if no blocks are available.
+    pub fn acquire(set: *BlockFreeSet) ?u64 {
+        const block = blk: {
+            if (set.recent.pop()) |block| {
+                break :blk block;
+            } else if (set.index.findFirstSet()) |shard| {
+                break :blk set.find_free_block_in_shard(shard) orelse unreachable;
+            } else return null;
+        };
+        const shard = block / shard_size;
+        assert(set.blocks.isSet(block));
+        assert(set.index.isSet(shard));
```

Comment:
```suggestion
        assert(set.index.isSet(shard));
        assert(set.blocks.isSet(block));
```

Swapping the order here so that we follow the same hierarchical order in which we use these as per our logic.